### PR TITLE
Fix mmkv id which contains special characters

### LINKED ISSUE #60: MMKV.mmkvWithID crash
```kotlin
MMKV.mmkvWithID("test/hello2").encode("hello", "world")
```

id中使用`/`会导致崩溃

### PR DESCRIPTION
**Pull request by zhongwuzw**
Fixes #60.

### CODE REVIEW

**Comment by lingol** [comment]
Now thinking about it, it simpler to store in MD5.

Let me explain. Replacing any special char with only `_` still has collisions. Say we have `test/hello2` and `test\hello2`, they all transform into `test_hello2`. 

While it's possible to keep a one-to-one mapping, like `/` -> `_` and `\` -> `=` and so on. There's still some collisions. Like `test/hello/2` and `test_hello/2`, all transform into `test_hello_2`.

So why not store in MD5, what do you think?

**Comment by zhongwuzw** [comment]
@lingol, oh yeah, seems we can't just simply replace the special character. 

`md5` is a substitute, but if we calculate value of `md5` in `C++`, we need to import a `md5` library? seems wasteful IMO. Another way is calculate `md5` in `Java`, and pass it to `C++`, but this need change the interface.

I have an idea, in `C++`, if we check `id` has any special character, we transform `id` to hex, for example, if `id` is `abc/`, we transform it to `"6162632f"`, just do a simply hex transform. What's your opinion?

And besides this, one more thing we need to consider, we use `id`  to be a filename, and we don't limit the length of `id`, but filesystem has path and file name length limits, should we need to consider this?

**Comment by lingol** [comment]
I would suggest using MD5, which has a fixed length. 
Passing MD5 from Java is very inconvenience. And in iOS there's no MD5 calculation in the OS level.
Given that we already import some of OpenSSL's code inside MMKV, there's little harm to import more.

So lets import OpenSSL's MD5 implementation.

**Comment by zhongwuzw** [comment]
@lingol 👌 , let's move to `md5`, I already extract the implementation of `md5` from `OpenSSL`, so I can just commit it or you need to import it？

**Comment by zhongwuzw** [comment]
These 5 files should be contained:
image

**Comment by lingol** [comment]
Wow, that's impressive. Sure you can go ahead with your extraction.

File: Android/MMKV/mmkv/src/main/cpp/MMKV.cpp [hunk 1]
```diff
@@ -1213,14 +1217,27 @@ static void mkSpecialCharacterFileDirectory() {
...
+}
+
 static string encodeFilePath(const string &mmapID) {
 	const char *specialCharacters = "\\/:*?\"<>|";
 	string filePath(mmapID);
 	bool hasSpecialCharacter = false;
 	for (int i = 0; i < filePath.size(); i++) {
 		if (strchr(specialCharacters, filePath[i]) != NULL) {
-			filePath[i] = '_';
+            md5(filePath);
```
Comment (lingol) [inline Android/MMKV/mmkv/src/main/cpp/MMKV.cpp]: Indentation should keep in sync.

File: Android/MMKV/mmkv/src/main/cpp/MMKV.cpp [hunk 2]
```diff
@@ -1213,14 +1217,27 @@ static void mkSpecialCharacterFileDirectory() {
...
+    {
+        sprintf(tmp, "%2.2x", md[i]);
+        strcat(buf, tmp);
+    }
+    value = buf;
+}
+
 static string encodeFilePath(const string &mmapID) {
 	const char *specialCharacters = "\\/:*?\"<>|";
 	string filePath(mmapID);
```
Comment (lingol) [inline Android/MMKV/mmkv/src/main/cpp/MMKV.cpp]: This string copy in unnecessary. I suggest making `void md5(string&)` into `string md5(const string&)`.

File: Android/MMKV/mmkv/src/main/cpp/MMKV.cpp [hunk 3]
```diff
@@ -1204,11 +1211,48 @@ bool MMKV::isFileValid(const std::string &mmapID) {
...
+    {
+        sprintf(tmp, "%2.2x", md[i]);
+        strcat(buf, tmp);
+    }
+    value = buf;
+}
+
+static string encodeFilePath(const string &mmapID) {
+    const char *specialCharacters = "\\/:*?\"<>|";
+    string filePath(mmapID);
```
Comment (lingol) [inline Android/MMKV/mmkv/src/main/cpp/MMKV.cpp]: This string copy is unnecessary. I suggest making void md5(string&) into string md5(const string&). And filePath = md5(mmapID).

Comment (zhongwuzw) [reply Android/MMKV/mmkv/src/main/cpp/MMKV.cpp]: @lingol 👍 I removed copy of constructor.

**Pull request merged**